File: app/src/main/java/eu/indiewalkabout/fridgemanager/ui/FoodListActivity.java
package eu.indiewalkabout.fridgemanager.ui;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.Observer;
import android.arch.lifecycle.ViewModelProviders;
import android.content.Intent;
import android.os.Build;
import android.support.annotation.Nullable;
import android.support.design.widget.FloatingActionButton;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.support.v7.widget.DividerItemDecoration;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.view.MenuItem;
import android.view.View;
import android.view.WindowManager;
import android.widget.TextView;
import android.widget.Toast;

import com.google.android.gms.ads.AdListener;
import com.google.android.gms.ads.AdView;
import com.google.android.gms.ads.InterstitialAd;
import com.hlab.fabrevealmenu.enums.Direction;
import com.hlab.fabrevealmenu.listeners.OnFABMenuSelectedListener;
import com.hlab.fabrevealmenu.view.FABRevealMenu;


import java.util.List;

import eu.indiewalkabout.fridgemanager.R;
import eu.indiewalkabout.fridgemanager.data.FoodEntry;

import static android.support.v7.widget.DividerItemDecoration.VERTICAL;


import eu.indiewalkabout.fridgemanager.util.ConsentSDK;
import eu.indiewalkabout.fridgemanager.util.GenericUtility;


/**
* ---------------------------------------------------------------------------------------------
* FoodListActivity
* Show list of food depending on FOOD_TYPE
* ---------------------------------------------------------------------------------------------
*/
public class FoodListActivity extends AppCompatActivity
        implements FoodListAdapter.ItemClickListener, OnFABMenuSelectedListener {

    public static final String TAG = FoodListActivity.class.getSimpleName();

    private Toolbar         foodsListToolbar;
    private RecyclerView    foodList;
    private FoodListAdapter foodListAdapter;
    private TextView        emptyListText, toolbarTitle;
    private FABRevealMenu   fabMenu;
    private FloatingActionButton fab ;


    private AdView mAdView;
    private InterstitialAd mInterstitialAd;

    // Key constant to use as key for intent extra
    // get the type of list to show from intent content extra
    public static final String FOOD_TYPE           = "food_type";

    // Values to use for passing intent extras in key/value pair
    // expiring food flag
    public static final String FOOD_EXPIRING       = "ExpiringFood";

    // expiring food Today flag
    public static final String FOOD_EXPIRING_TODAY = "ExpiringFoodToday";

    // consumed food flag
    public static final String FOOD_SAVED          = "SavedFood";

    // expired food flag
    public static final String FOOD_DEAD           = "DeadFood";

    // Hold the type of food to show in list, use here and by adapter
    // TODO : find a better way to pass this info to adapter
    private String foodlistType;





    /**
    * ---------------------------------------------------------------------------------------------
    * onCreate
    * @param savedInstanceState
    * ---------------------------------------------------------------------------------------------
    */
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_food_list);

        mAdView = findViewById(R.id.adView);

        mAdView.loadAd(ConsentSDK.getAdRequest(FoodListActivity.this));

        // empty view for empty list message
        emptyListText = findViewById(R.id.empty_view);


        // TODO : use this in another way
        // get intent extra for configuring list type
        Intent intent = getIntent();
        if (intent != null && intent.hasExtra(FOOD_TYPE)) {
            foodlistType = intent.getStringExtra(FOOD_TYPE);
            Log.d(TAG, "onCreate: FOOD_TYPE : " + foodlistType.toString());
        }

        // init toolbar
        toolBarInit();

        // navigation fab btn
        addRevealFabBtn();

        // init recycle view list
        initRecycleView();

        // make bottom navigation bar and status bar hide
        hideStatusNavBars();
    }



    /**
     * ---------------------------------------------------------------------------------------------
     * Make bottom navigation bar and status bar hide, without resize when reappearing
     * ---------------------------------------------------------------------------------------------
     */
    private void hideStatusNavBars() {
        // minsdk version is 19, no need code for lower api
        View decorView = getWindow().getDecorView();

        // hide status bar
        if (Build.VERSION.SDK_INT < 16) {
            getWindow().setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN,
                    WindowManager.LayoutParams.FLAG_FULLSCREEN);
        } else if (Build.VERSION.SDK_INT >= 16){
            decorView.setSystemUiVisibility(View.SYSTEM_UI_FLAG_FULLSCREEN);
        }

        // hide navigation bar
        if(Build.VERSION.SDK_INT > 11 && Build.VERSION.SDK_INT < 19) {
            View v = this.getWindow().getDecorView();
            v.setSystemUiVisibility(View.GONE);
        } else if(Build.VERSION.SDK_INT >= 19) {
            int uiOptions = View.SYSTEM_UI_FLAG_HIDE_NAVIGATION |
                    View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY;
            decorView.setSystemUiVisibility(uiOptions);
        }
    }



    // Recycle touch an item callback to update/modify task
    @Override
    public void onItemClickListener(int itemId) {
        Log.d(TAG, "onItemClickListener: Item" + itemId + "touched.");
    }




    /**
    * ---------------------------------------------------------------------------------------------
    * Toolbar init
    * ---------------------------------------------------------------------------------------------
    */
    private void toolBarInit() {
        // get the toolbar
        foodsListToolbar = (Toolbar) findViewById(R.id.food_list_toolbar);

        setToolBarTitle();

        // place toolbar in place of action bar
        setSupportActionBar(foodsListToolbar);

        // get a support action bar
        ActionBar actionBar = getSupportActionBar();

        // up button
        actionBar.setDisplayHomeAsUpEnabled(true);

    }


    /**
     * ---------------------------------------------------------------------------------------------
     * Set Toolbar title
     * ---------------------------------------------------------------------------------------------
     */
    private void setToolBarTitle() {
        toolbarTitle = foodsListToolbar.findViewById(R.id.toolbar_title_tv);

        // set correct title
        if (foodlistType.equals(FOOD_EXPIRING)) {
            // foodsListToolbar.setTitle(R.string.foodExpiring_activity_title);
            toolbarTitle.setText(R.string.foodExpiring_activity_title);
            Log.d(TAG, "onCreate: FOOD_TYPE : " + R.string.foodExpiring_activity_title);

        } else if (foodlistType.equals(FOOD_SAVED)) {
            // foodsListToolbar.setTitle(R.string.R.string.foodSaved_activity_title);
            toolbarTitle.setText(R.string.foodSaved_activity_title);
            Log.d(TAG, "onCreate: FOOD_TYPE : " + R.string.foodSaved_activity_title);

        } else if (foodlistType.equals(FOOD_DEAD)) {
            // foodsListToolbar.setTitle(R.string.foodDead_activity_title);
            toolbarTitle.setText(R.string.foodDead_activity_title);
            Log.d(TAG, "onCreate: FOOD_TYPE : " + R.string.foodDead_activity_title);

        }
    }


    /**
     * ---------------------------------------------------------------------------------------------
     * Show an interstitial on ui request
     * ---------------------------------------------------------------------------------------------
     */
    private void showInterstitialAd() {

        mInterstitialAd = new InterstitialAd(this);
        mInterstitialAd.setAdUnitId(getResources().getString(R.string.admob_key_interstitial));

        mInterstitialAd.loadAd(ConsentSDK.getAdRequest(this));
        mInterstitialAd.setAdListener(new AdListener() {
            @Override
            public void onAdLoaded() {
                // for debug :
                // Toast.makeText(FoodListActivity.this, "Ad Interstitial LOADED", Toast.LENGTH_SHORT).show();
                // Show interstitial
                mInterstitialAd.show();
                super.onAdLoaded();
            }

            @Override
            public void onAdFailedToLoad(int errorCode) {
                // for debug :
                // Toast.makeText(FoodListActivity.this, "Ad Interstitial   FAILED TO LOAD", Toast.LENGTH_SHORT).show();
                // Log.i(TAG, "onAdFailedToLoad: ConsentSDK.getAdRequest(MainActivity.this) : "
                //        +ConsentSDK.getAdRequest(MainActivity.this));
            }
        });

    }

    /**
    * ---------------------------------------------------------------------------------------------
    * Recycle view list init
    * ---------------------------------------------------------------------------------------------
    */
    private void initRecycleView(){
        foodList = findViewById(R.id.food_list_recycleView);

        if (foodList == null) {
            Log.d(TAG, "onCreate: foodList == null ");
        }

        foodList.setLayoutManager(new LinearLayoutManager(this));


        // Init adapter
        foodListAdapter = new FoodListAdapter(this, this, foodlistType);
        foodList.setAdapter(foodListAdapter);

        // Divider
        DividerItemDecoration decoration = new DividerItemDecoration(getApplicationContext(), VERTICAL);
        foodList.addItemDecoration(decoration);

        // Configure the adpater; it uses LiveData to keep updated on changes
        setupAdapter();

        // make fab button hide when scrolling list
        foodList.addOnScrollListener(new RecyclerView.OnScrollListener() {

            @Override
            public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
                if (dy > 0 || dy < 0 && fab.isShown())
                    fab.hide();
            }

            @Override
            public void onScrollStateChanged(RecyclerView recyclerView, int newState) {
                if (newState == RecyclerView.SCROLL_STATE_IDLE)
                    fab.show();
                super.onScrollStateChanged(recyclerView, newState);
            }
        });

    }






    /**
     * ---------------------------------------------------------------------------------------------
     * Used to reload from db the tasks list and update the list view in screen
     * ---------------------------------------------------------------------------------------------
     */
    private void setupAdapter() {

        Log.d(TAG, "setupAdapter: LOAD FOOD ENTRIES IN LIST ");
        retrieveAllFood();
        Log.d(TAG, "setupAdapter: FOOD_TYPE : " + foodlistType);


    }



    /**
     * ---------------------------------------------------------------------------------------------
     * Livedata/ViewModel recovering Expiring Food list
     * ---------------------------------------------------------------------------------------------
     */
    private void retrieveAllFood() {
        Log.d(TAG, "Actively retrieving Expiring Food from DB");

        // Instance factory FoodsViewModelFactory, parametrized with foodlistType
        FoodsViewModelFactory factory = new FoodsViewModelFactory(foodlistType);

        // Create the FoodsViewModel  for the food list, based on  foodlistType
        final FoodsViewModel viewModel = ViewModelProviders.of(this,factory).get(FoodsViewModel.class);

        // Observe changes in data through LiveData: getFoodList() actually return LiveData<List<FoodEntry>>
        LiveData<List<FoodEntry>> foods = viewModel.getFoodList();
        foods.observe(this, new Observer<List<FoodEntry>>() {
            @Override
            public void onChanged(@Nullable List<FoodEntry> foodEntries) {
                Log.d(TAG, "Receiving database "+ foodlistType + " LiveData");
                foodListAdapter.setFoodEntries(foodEntries);
                if (foodEntries.size() > 0 ) {
                    emptyListText.setVisibility(View.INVISIBLE);
                } else {
                    emptyListText.setVisibility(View.VISIBLE);
                }
            }
        });

    }





    /**
    * ----------------------------------------------------------------------------------------------
    *                                          MENU STUFF
    * ----------------------------------------------------------------------------------------------
    */
    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();

        // When the home button is pressed, take the user back to Home
        if (id == android.R.id.home) {
            // show interstitial ad on back home only 50% of times
            int guess = GenericUtility.randRange_ApiCheck(1,10);
            if (guess <=4) {
                showInterstitialAd();
            }

            // showInterstitialAd();
            onBackPressed();
        }

        return super.onOptionsItemSelected(item);
    }


    // ---------------------------------------------------------------------------------------------
    //                                  REVEALING FAB BTN STUFF
    // ---------------------------------------------------------------------------------------------
    public void onBackPressed() {
        super.onBackPressed();
        if (fabMenu != null) {
            if (fabMenu.isShowing()) {
                fabMenu.closeMenu();

            }
        }
    }

    public FABRevealMenu getFabMenu() {
        return fabMenu;
    }

    public void setFabMenu(FABRevealMenu fabMenu) {
        this.fabMenu = fabMenu;
    }

    /**
     * ---------------------------------------------------------------------------------------------
     * Adding revealing main_fab button
     * ---------------------------------------------------------------------------------------------
     */
    private void addRevealFabBtn(){
        fab     = findViewById(R.id.food_list_fab);
        fabMenu = findViewById(R.id.food_list_fabMenu);

        // choose what menu item must be seen based on type of list
        if (foodlistType.equals(FoodListActivity.FOOD_EXPIRING)){
            fabMenu.removeItem(R.id.menu_expiring_food);

        } else if (foodlistType.equals(FoodListActivity.FOOD_DEAD)){
            fabMenu.removeItem(R.id.menu_dead_food);

        } else if (foodlistType.equals(FoodListActivity.FOOD_SAVED)){
            fabMenu.removeItem(R.id.menu_consumed_food);
        }

        try {
            if (fab != null && fabMenu != null) {
                setFabMenu(fabMenu);

                //attach menu to main_fab
                fabMenu.bindAnchorView(fab);

                //set menu selection listener
                fabMenu.setOnFABMenuSelectedListener(this);
            }
        } catch (Exception e) {
            e.printStackTrace();
        }

        fabMenu.setMenuDirection(Direction.LEFT);

    }


    /**
     * ---------------------------------------------------------------------------------------------
     * Revealing main_fab button menu management
     * ---------------------------------------------------------------------------------------------
     */
    @Override
    public void onMenuItemSelected(View view, int id) {
        if (id == R.id.menu_insert) {
            Intent toInsertFood = new Intent(FoodListActivity.this, InsertFoodActivity.class);
            startActivity(toInsertFood);

        } else if (id == R.id.menu_expiring_food) {
            Intent showExpiringFood = new Intent(FoodListActivity.this, FoodListActivity.class);
            showExpiringFood.putExtra(FoodListActivity.FOOD_TYPE, FoodListActivity.FOOD_EXPIRING);
            startActivity(showExpiringFood);

        } else if (id == R.id.menu_consumed_food) {
            int guess = GenericUtility.randRange_ApiCheck(1,10);
            if (guess <=3) {
                showInterstitialAd();
            }
            Intent showSavedFood = new Intent(FoodListActivity.this, FoodListActivity.class);
            showSavedFood.putExtra(FoodListActivity.FOOD_TYPE, FoodListActivity.FOOD_SAVED);
            startActivity(showSavedFood);

        } else if (id == R.id.menu_dead_food) {
            int guess = GenericUtility.randRange_ApiCheck(1,10);
            if (guess <=3) {
                showInterstitialAd();
            }
            Intent showDeadFood = new Intent(FoodListActivity.this, FoodListActivity.class);
            showDeadFood.putExtra(FoodListActivity.FOOD_TYPE, FoodListActivity.FOOD_DEAD);
            startActivity(showDeadFood);

        } else if (id == R.id.menu_home) {
            // show interstitial ad on back home only 50% of times

            int guess = GenericUtility.randRange_ApiCheck(1,10);
            if (guess <=4) {
                showInterstitialAd();
            }

            // showInterstitialAd();

            Intent returnHome = new Intent(FoodListActivity.this, MainActivity.class);
            startActivity(returnHome);
        }
    }

}

File: app/src/main/java/eu/indiewalkabout/fridgemanager/ui/FoodListActivity.kt
package eu.indiewalkabout.fridgemanager.ui

import androidx.lifecycle.LiveData
import androidx.lifecycle.Observer
import android.content.Intent
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import android.util.Log
import android.view.MenuItem
import android.view.View
import androidx.lifecycle.ViewModelProvider
import com.google.android.gms.ads.AdListener
import com.google.android.gms.ads.InterstitialAd
import com.hlab.fabrevealmenu.enums.Direction
import com.hlab.fabrevealmenu.listeners.OnFABMenuSelectedListener
import eu.indiewalkabout.fridgemanager.App
import eu.indiewalkabout.fridgemanager.R
import eu.indiewalkabout.fridgemanager.data.model.FoodEntry
import eu.indiewalkabout.fridgemanager.ui.FoodListAdapter.ItemClickListener
import eu.indiewalkabout.fridgemanager.util.ConsentSDK.Companion.getAdRequest
import eu.indiewalkabout.fridgemanager.util.GenericUtility.hideStatusNavBars
import eu.indiewalkabout.fridgemanager.util.GenericUtility.randRange_ApiCheck
import eu.indiewalkabout.fridgemanager.util.GenericUtility.showRandomizedInterstAds
import kotlinx.android.synthetic.main.activity_food_list.*
import kotlinx.android.synthetic.main.activity_food_list.adView
import kotlinx.android.synthetic.main.activity_food_list.emptyListText
import kotlinx.android.synthetic.main.activity_main.*
import java.text.DateFormat.getDateInstance
import java.util.*

// ---------------------------------------------------------------------------------------------
// Show list of food depending on FOOD_TYPE

class FoodListActivity : AppCompatActivity(), ItemClickListener, OnFABMenuSelectedListener {

    companion object {
        val TAG = FoodListActivity::class.java.simpleName

        // Key constant to use as key for intent extra
        // get the type of list to show from intent content extra
        const val FOOD_TYPE = "food_type"

        // Values to use for passing intent extras in key/value pair
        // expiring food flag
        const val FOOD_EXPIRING = "ExpiringFood"

        // expiring food Today flag
        const val FOOD_EXPIRING_TODAY = "ExpiringFoodToday"

        // saved food flag
        const val FOOD_SAVED = "SavedFood"

        // dead food flag
        const val FOOD_DEAD = "DeadFood"
    }

    var foodListAdapter: FoodListAdapter? = null
    var foodListForShare: String = ""
    var foodShareSubject: String = ""

    lateinit var mInterstitialAd: InterstitialAd

    // Hold the type of food to show in list, use here and by adapter
    // TODO : find a better way to pass this info to adapter
    private var foodlistType: String? = null



    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_food_list)

        // You have to pass the AdRequest from ConsentSDK.getAdRequest(this) because it handle the right way to load the ad
        adView.loadAd(getAdRequest(this@FoodListActivity))

        // TODO : use this in another way
        // get intent extra for configuring list type
        val intent = intent
        if (intent != null && intent.hasExtra(FOOD_TYPE)) {
            foodlistType = intent.getStringExtra(FOOD_TYPE)
            Log.d(TAG, "onCreate: FOOD_TYPE : " + foodlistType.toString())
        }

        toolBarInit()
        addRevealFabBtn()
        initRecycleView()

        hideStatusNavBars(this)

        share_food_list.setOnClickListener {
            val sharingIntent = Intent(Intent.ACTION_SEND)
            sharingIntent.type = "text/plain"
            val shareBody = foodListForShare
            sharingIntent.putExtra(Intent.EXTRA_SUBJECT, foodShareSubject)
            sharingIntent.putExtra(Intent.EXTRA_TEXT, shareBody)
            startActivity(Intent.createChooser(sharingIntent, "Share via"))
        }
    }


    // Recycle touch an item callback to update/modify task
    override fun onItemClickListener(itemId: Int) {
        Log.d(FoodListActivity.TAG, "onItemClickListener: Item" + itemId + "touched.")
    }

    // ---------------------------------------------------------------------------------------------
    // Toolbar init

    private fun toolBarInit() {
        setToolBarTitle()
        // place toolbar in place of action bar
        setSupportActionBar(food_list_toolbar)

        // up button
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)
    }


    private fun setToolBarTitle() {

        // set correct title
        if (foodlistType == FOOD_EXPIRING) {
            toolbar_title_tv.setText(R.string.foodExpiring_activity_title)
            foodShareSubject = getString(R.string.expiring_food_list_subject)
            Log.d(TAG, "onCreate: FOOD_TYPE : " + R.string.foodExpiring_activity_title)

        } else if (foodlistType == FOOD_SAVED) {
            toolbar_title_tv.setText(R.string.foodSaved_activity_title)
            foodShareSubject = getString(R.string.saved_food_list_subject)
            Log.d(TAG, "onCreate: FOOD_TYPE : " + R.string.foodSaved_activity_title)

        } else if (foodlistType == FOOD_DEAD) {
            toolbar_title_tv.setText(R.string.foodDead_activity_title)
            foodShareSubject = getString(R.string.wasted_food_list_subject)
            Log.d(TAG, "onCreate: FOOD_TYPE : " + R.string.foodDead_activity_title)
        }
    }

    // Show admob interstitial on ui request
    private fun showAdMobInterstitialAd() {
        mInterstitialAd = InterstitialAd(this)
        mInterstitialAd.adUnitId = resources.getString(R.string.admob_key_interstitial)

        // You have to pass the AdRequest from ConsentSDK.getAdRequest(this) because it handle the right way to load the ad
        mInterstitialAd.loadAd(getAdRequest(this))
        mInterstitialAd.adListener = object : AdListener() {
            override fun onAdLoaded() {
                // Toast.makeText(FoodListActivity.this, "Ad Interstitial LOADED", Toast.LENGTH_SHORT).show();
                // Show interstitial
                mInterstitialAd.show()
                super.onAdLoaded()
            }

            override fun onAdFailedToLoad(errorCode: Int) {
                // Code to be executed when an ad request fails.
                // Toast.makeText(FoodListActivity.this, "Ad Interstitial   FAILED TO LOAD", Toast.LENGTH_SHORT).show();
                // Log.i(TAG, "onAdFailedToLoad: ConsentSDK.getAdRequest(MainActivity.this) : "
                //        +ConsentSDK.getAdRequest(MainActivity.this));
            }
        }
    }


    private fun initRecycleView() {

        if (food_list_recycleView == null) {
            Log.d(TAG, "onCreate: foodList == null ")
        }

        food_list_recycleView.setLayoutManager(LinearLayoutManager(this))

        foodListAdapter = FoodListAdapter(this, this, foodlistType!!)
        food_list_recycleView.setAdapter(foodListAdapter)

        setupAdapter()

        // make fab button hide when scrolling list
        food_list_recycleView.addOnScrollListener(object : androidx.recyclerview.widget.RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: androidx.recyclerview.widget.RecyclerView, dx: Int, dy: Int) {
                if (dy > 0 || dy < 0 && food_list_fab!!.isShown) food_list_fab!!.hide()
            }

            override fun onScrollStateChanged(recyclerView: androidx.recyclerview.widget.RecyclerView, newState: Int) {
                if (newState == androidx.recyclerview.widget.RecyclerView.SCROLL_STATE_IDLE) food_list_fab!!.show()
                super.onScrollStateChanged(recyclerView, newState)
            }
        })
    }

    // Used to reload from db the tasks list and update the list view in screen
    private fun setupAdapter() {
        Log.d(TAG, "setupAdapter: LOAD FOOD ENTRIES IN LIST ")
        retrieveAllFood()
        Log.d(TAG, "setupAdapter: FOOD_TYPE : $foodlistType")
    }


    private fun retrieveAllFood() {
        Log.d(TAG, "Actively retrieving Expiring Food from DB")

        val factory = FoodsViewModelFactory(foodlistType!!)

        // Create the FoodsViewModel  for the food list, based on  foodlistType
        val viewModel = ViewModelProvider(this, factory).get(FoodsViewModel::class.java)

        // Observe changes in data through LiveData: getFoodList() actually return LiveData<List<FoodEntry>>
        val foods: LiveData<MutableList<FoodEntry>>? = viewModel.foodList
        foods!!.observe(this, Observer<MutableList<FoodEntry>?> { foodEntries ->
            Log.d(TAG, "Receiving database $foodlistType LiveData")
            foodListAdapter!!.setFoodEntries(foodEntries)
            if (foodEntries!!.size > 0) {
                emptyListText.visibility = View.INVISIBLE
                share_food_list.visibility = View.VISIBLE
                // fill the list for sharing
                foodListForShare += "${foodShareSubject} \n\n"
                for(item in foodEntries) {
                    foodListForShare += "${item.name} :  ${getDateInstance().format(item.expiringAt)}\n\n"
                }
            } else {
                emptyListText.visibility = View.VISIBLE
                share_food_list.visibility = View.INVISIBLE
            }
        })
    }

    // MENU STUFF
    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId

        // When the home button is pressed, take the user back to Home
        if (id == android.R.id.home) {
            showRandomizedInterstAds(4, this)
            onBackPressed()
        }
        return super.onOptionsItemSelected(item)
    }


    // ---------------------------------------------------------------------------------------------
    //                                  REVEALING FAB BTN STUFF
    // ---------------------------------------------------------------------------------------------
    override fun onBackPressed() {
        super.onBackPressed()
        showRandomizedInterstAds(4, this)
        if (food_list_fabMenu != null) {
            if (food_list_fabMenu!!.isShowing) {
                food_list_fabMenu!!.closeMenu()
            }
        }
    }

    /**
     * ---------------------------------------------------------------------------------------------
     * Adding revealing main_fab button
     * ---------------------------------------------------------------------------------------------
     */
    private fun addRevealFabBtn() {

        // choose what menu item must be seen based on type of list
        if (foodlistType == FOOD_EXPIRING) {
            food_list_fabMenu?.removeItem(R.id.menu_expiring_food)

        } else if (foodlistType == FOOD_DEAD) {
            food_list_fabMenu?.removeItem(R.id.menu_dead_food)

        } else if (foodlistType == FOOD_SAVED) {
            food_list_fabMenu?.removeItem(R.id.menu_consumed_food)
        }
        try {
            if (food_list_fab != null && food_list_fabMenu != null) {

                //attach menu to main_fab
                food_list_fabMenu!!.bindAnchorView(food_list_fab!!)

                //set menu selection listener
                food_list_fabMenu!!.setOnFABMenuSelectedListener(this)
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
        food_list_fabMenu?.setMenuDirection(Direction.LEFT)
    }

    /**
     * ---------------------------------------------------------------------------------------------
     * Revealing main_fab button menu management
     * ---------------------------------------------------------------------------------------------
     */
    override fun onMenuItemSelected(view: View, id: Int) {
        if (id == R.id.menu_insert) {
            val toInsertFood = Intent(this@FoodListActivity, InsertFoodActivity::class.java)
            startActivity(toInsertFood)

        } else if (id == R.id.menu_expiring_food) {
            showExpiringFood()

        } else if (id == R.id.menu_consumed_food) {
            showRandomizedInterstAds(3, this)
            showSavedFood()

        } else if (id == R.id.menu_dead_food) {
            showRandomizedInterstAds(3, this)
            showDeadFood()

        } else if (id == R.id.menu_home) {
            showRandomizedInterstAds(4, this)
            returnHome()
        }
    }

    private fun returnHome() {
        val returnHome = Intent(this@FoodListActivity, MainActivity::class.java)
        startActivity(returnHome)
    }


    private fun showDeadFood() {
        val showDeadFood = Intent(this@FoodListActivity, FoodListActivity::class.java)
        showDeadFood.putExtra(FOOD_TYPE, FOOD_DEAD)
        startActivity(showDeadFood)
    }

    private fun showSavedFood() {
        val showSavedFood = Intent(this@FoodListActivity, FoodListActivity::class.java)
        showSavedFood.putExtra(FOOD_TYPE, FOOD_SAVED)
        startActivity(showSavedFood)
    }

    private fun showExpiringFood() {
        val showExpiringFood = Intent(this@FoodListActivity, FoodListActivity::class.java)
        showExpiringFood.putExtra(FOOD_TYPE, FOOD_EXPIRING)
        startActivity(showExpiringFood)
    }





}